src/subsplit_dag_storage.hpp

```diff
@@ -378,23 +378,84 @@ class GenericVerticesView {
 using VerticesView = GenericVerticesView<SubsplitDAGStorage>;
 using ConstVerticesView = GenericVerticesView<const SubsplitDAGStorage>;
 
+template <typename T>
+class PrependableVector {
+public:
+  explicit PrependableVector(PrependableVector<T>* prepend = nullptr) : prepend_{prepend ? &prepend->data_ : nullptr} {}
+
+  T& at(size_t i) {
+    if (!prepend_) return data_.at(i);
+    if (i < prepend_->size()) return (*prepend_)[i];
+    return data_.at(i - prepend_->size());
+  }
+
+  const T& at(size_t i) const {
+    if (!prepend_) return data_.at(i);
+    if (i < prepend_->size()) return (*prepend_)[i];
+    return data_.at(i - prepend_->size());
+  }
+
+  T& operator[](size_t i) {
+    if (!prepend_) return data_[i];
+    if (i < prepend_->size()) return (*prepend_)[i];
+    return data_[i - prepend_->size()];
+  }
+
+  const T& operator[](size_t i) const {
+    if (!prepend_) return data_[i];
+    if (i < prepend_->size()) return (*prepend_)[i];
+    return data_[i - prepend_->size()];
+  }
+
+  size_t size() const {
+    if (!prepend_) return data_.size();
+    return prepend_->size() + data_.size();
+  }
+
+  void resize(size_t new_size) {
+    if (!prepend_) data_.resize(new_size);
+    else data_.resize(new_size - prepend_->size());
+  }
+
+  PrependableVector& operator=(const std::vector<T>& data) {
+    data_ = data;
+    return *this;
+  }
+
+  bool HavePrepend() const {
+    return prepend_;
+  }
+
+  size_t PrependSize() const {
+    if (!prepend_) return 0;
+    return prepend_->size();
+  }
+
+private:
+  std::vector<T> data_;
+  std::vector<T>* prepend_;
+};
+
+struct PrependTag {};
```

Comment:
It's a common technique called "tag dispatching", but I can change terminology to "hint" for example. Without tag dispatching the constructor looks a lot like the copy constructor, and I'll have to define copy ctors and assignment operators to SubsplitDAG and all classes that derives from it. This would be bug prone as well as verbose.

Reply:
Cool to learn about this, and got it!

Perhaps `PrependDispatchTag` with a little comment and perhaps a link to a description of this idiom?